### PR TITLE
WIP: gobject-introspection: Support cross

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

The idea is to build the python tools separate from the C bits, and the python tools are run natively (including C extension module, in particular), but the C bits are either the installed library, or tools that must be run with an emulator (during cross compilation) and in both cases are non-native.

I need to still need to wrap the two together for downstream consumption, and I am also debating exactly how I'd like to do that.

CC @jtojnar @samueldr

Upstream issue: https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/337
Upstream PR: https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/224

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
Where do docs fit into this #87904?

**Comment by Ericson2314** [comment]
@jtojnar Those look perfectly compatible for me :), We could build them with the python stuff or the C stuff. If you merge your PR I'll be happy to rebase mine on top.

**Comment by matthewbauer** [comment]
Do all dependents of gobject-introspection need exe_wrapper?

When cross compiling nixos, I ended up having to disable gobject-introspection, vala, and gtk-docs. It ends up working pretty well, since you're already building everything from scratch, and I've never encountered a point where I need any of the 3 (https://github.com/matthewbauer/nixpkgs/commit/f33e8c3d6e1c7494755868862e26e3809a394479). Annoyingly there are a few things that you have to patch to disable introspection correctly.

**Comment by Ericson2314** [comment]
The exe wrapper is sadly mandatory for anything generating typelibs (compiled gir files), but anything purely consuming them at runtime doesn't need the exe wrapper: it just links `girepository` like any other C library.

I would like to split up the package (different wrappers around specific outputs from the two underlying builds) to reflect this.

**Comment by matthewbauer** [comment]
> The exe wrapper is sadly mandatory for anything generating typelibs (compiled gir files), but anything purely consuming them at runtime doesn't need the exe wrapper: it just links girepository like any other C library.

Well I would be kind of surprised if you needed the emulator at runtime - it would be the cross compiled architecture :)

Meson could just provide exe_wrapper everywhere for consistency. qemu-user at least has few dependencies.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -11947,11 +11947,22 @@ in
   gns3-gui = gns3Packages.guiStable;
   gns3-server = gns3Packages.serverStable;
 
-  gobject-introspection = callPackage ../development/libraries/gobject-introspection {
+  gobject-introspection-full-classic = callPackage ../development/libraries/gobject-introspection {
     nixStoreDir = config.nix.storeDir or builtins.storeDir;
     inherit (darwin) cctools;
   };
 
+  gobject-introspection-py-tools = gobject-introspection-full-classic.override {
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:11955]: Can you explain the when to use what?

Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix:11955]: Well I think I'll need to make a wrapper package so this isn't actually used directly. But the idea is (whether directly or via some wrapper) than the py tools are need when generating new gir files and type libs when building some C library, and aren't needed by the programs in python etc that just consume typelibs.

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix:11955]: I don’t think we should make that distinction if it’s not made by gobject introspection. IIUC things that consume gir also produce gir without much distinction.

Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix:11955]: That's fine, but we will need the wrapper anyways to combine the build platform py tools with host platform everything else.

**Comment by Atemu** [comment]
What's the status on this?

One of the patches was merged a few months ago.

**Comment by Artturin** [comment]
rebased in https://github.com/Artturin/nixpkgs/tree/gobject-introspectcross
left the source at the normal gi src


during build it fails with 
```
gobject-introspection> substitute(): ERROR: file '/nix/store/zq92yh2l39qnjmjdn6zmiyav9ai6yb7i-gobject-introspection-hook'does not exist
```

> What's the status on this?
> 
> One of the patches was merged a few months ago.

need this for tests https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/224

**Comment by SuperSandro2000** [comment]
Do we actually need the patch at all? Could we just build everything both times and throw away what we don't need? That would be a bit wasteful but maintaining the patch is way to much work.

**Pull request closed without merging**